### PR TITLE
feature/pfdetect-regex-parser

### PR DESCRIPTION
**Pull request by jrouzierinverse**
Description
-----------
A generic parser for defining actions for syslog entries

Impacts
-----------
pfdetect

NEWS file entries
-------

New Features
++++++++++++
* A new generic pfdetect parser for custom actions

Bug Fixes
+++++++++

Delete branch after merge
-------------------------
NO

Upgrade file entries
-------------------------
Run the following script to fix the pfdetect.conf
```
/usr/local/pf/addons/upgrade/to-6.5-pfdetect-conf.pl
```

### CODE REVIEW

**Comment by jrouzierinverse** [comment]
This is ready for review

**Comment by jrouzierinverse** [comment]
Please review me

**Comment by jrouzierinverse** [comment]
What still needs to be done is was which Api Actions are going to be allowed to be displayed in the choice list.

@fdurand is this something you can help me with?

**Comment by julsemaan** [comment]
TBH reviewing a PR of 118 commits containing 2000+ lines doesn't really seem like something that should be for 6.5 with the release window we have

I'll leave the call to @extrafu but considering the huge amount of code there is, it doesn't look like it can be reviewed+tested correctly in the limited amount of time we have, especially considering there is still a missing piece in the PR and the end of the merge window was a few hours ago

There's nothing stopping us from having a small 6.6 with this being the main feature in 4-5 weeks

image

**Comment by extrafu** [comment]
Please @fdurand and @julsemaan give it a try and we'll advise after.

**Comment by julsemaan** [comment]
What I can see missing now (didn't fully look at the code yet):
- Review+adjust documentation of admin guide for syslog parsers creation (like suricata MD5, security onion, ...)
- Documentation on new regex parsers and how to use them (especially for the actions)

Will test it out, but I still maintain my point of this being too last minute. Because even if the code works it doesn't mean it can be reviewed correctly.

**Review by julsemaan — COMMENTED** [review]
The detect parsers unit test are broken because of the logic change in the factory

**Review by julsemaan — CHANGES_REQUESTED** [review]
for now found:
- missing doc
- broken unit tests

**Comment by julsemaan** [comment]
Suricata/snort parser is working as expected in an integration test

**Review by julsemaan — CHANGES_REQUESTED** [review]
Suricata parser shows as disabled in the admin, but the pfdetect process is started and working

Note that this is an existing parser I had and not one I created while testing

conf:
[suricata]
type=suricata
path=/usr/local/pf/var/suricata_alert

image

File: lib/pf/detect/parser.pm [hunk 1]
```diff
@@ -16,7 +16,15 @@ Base class for a pfdetect parser
...
 use warnings;
 use Moo;
+
+has id => (is => 'rw', required => 1);
+
+has path => (is => 'rw', required => 1);
+
+has type => (is => 'rw', required => 1);
  
+has enabled => (is => 'rw', default =>  sub { "enabled" });
```
Comment (julsemaan) [inline lib/pf/detect/parser.pm]: Same semantic logic as in pfmon, we should use status and not enabled as the attribute name

**Review by julsemaan — CHANGES_REQUESTED** [review]
When creating a new regex syslog parser, clicking save doesn't do a thing (Firefox on Ubuntu 16)

Config I have:
image

EDIT: I found the issue. Its that nothing was defined in the rule and the mandatory fields in it were required. Nothing indicated were the issue was. Should there be an issue with a rule, it should expand the rule so we can see the error

**Review by julsemaan — CHANGES_REQUESTED** [review]
Success/error should display below the navigation bar like in other sections

image

**Review by julsemaan — CHANGES_REQUESTED** [review]
The help contains a lorem ipsum instead of real help

**Comment by julsemaan** [comment]
I'll stop here for now and let @jrouzierinverse review the requested changes and we'll continue after

With the amount of new UI and the alpha stage were that code is, I don't really consider this as 6.5 material as we'll have to spend quite a bit of time testing + adjusting this instead of focusing on the already big amount of features that have been merged in 6.5 which also require to be tested

**Comment by fdurand** [comment]
When i try to create a dhcp parser.

 httpd.admin(11338) ERROR: Caught exception in pfappserver::Controller::Config::Pfdetect->view "Can't call method "process" on an undefined value at /usr/local/pf/html/pfappserver/lib/pfappserver/Base/Controller/Crud.pm line 206." (pfappserver::PacketFence::Controller::Root::end)

**Comment by jrouzierinverse** [comment]
Fixed
* Tests 
* Creating a dhcp parser
* Poddoc
* Upgrade instructions
* Expand accordions on error.
* DIsplay of success/error message

Updating the documentation is the next thing to complete.

**Comment by extrafu** [comment]
What is the status of the PR? We need to merge it ASAP to begin testing.

**Comment by julsemaan** [comment]
Won't re-read the whole code so I'd say this is fairly ready.

As long as @jrouzierinverse reviewed the admin guide to make sure syslog parser sections are still relevant for this new PR, then this is ready for testing/merging

**Review by julsemaan — APPROVED** [review]

**Comment by jrouzierinverse** [comment]
Check documentation update please

**Comment by extrafu** [comment]
@julsemaan said ".. As long as @jrouzierinverse reviewed the admin guide ... " - have you done so?

**Comment by extrafu** [comment]
@jrouzierinverse You haven't answered my question before merging.

**Comment by jrouzierinverse** [comment]
Sorry

**Pull request merged**